From chirag@linux.vnet.ibm.com Mon Sep  8 03:50:06 2008
Date: Mon, 8 Sep 2008 13:07:32 +0530
From: Chirag Jog <chirag@linux.vnet.ibm.com>
To: xavier droubay <xavier.droubay@gmail.com>
Cc: linux-rt-users@vger.kernel.org, LKML <linux-kernel@vger.kernel.org>, Steven Rostedt <rostedt@goodmis.org>, Thomas Gleixner <tglx@linutronix.de>
Subject: [PATCH] Convert tlbstate_lock spin_lock to raw_spin_lock

    [ The following text is in the "iso-8859-1" character set. ]
    [ Your display is set for the "ANSI_X3.4-1968" character set.  ]
    [ Some characters may be displayed incorrectly. ]

* xavier droubay <xavier.droubay@gmail.com> [2008-09-08 03:06:51]:

> Hello,
> 
> I experienced this "rtmutex.c:743" known bug with rt3 patch, using
> 
>  http://www.kernel.org/pub/linux/kernel/v2.6/linux-2.6.26.3.tar.bz2
>  http://rt.et.redhat.com/download/patch-2.6.26.3-rt3.bz2
> 
> I compiled this kernel with the "make-kpkg" debian utility.
> 
> I first compiled this kernel with NO SMP, NO HIGHMEM, and partial ACPI
> (without fan, processor, ... modules). Attached file
> "config-2.6.26.3-rt3lowlty.frwr.gz".
> I didn't experience any problem with this kernel.
> 
> I then enabled SMP and HIGHMEM, and experienced  "rtmutex.c:743" BUG.
> Attached files
> "config-2.6.26.3-rt3rt-smp-hm.gz" and  "dmesg.gz" for some BUGGY message.
> 
> 
> I have seen somewhere on the archive, that some people still
> experienced problems with SMP.
> I guess no fix has been done from rt3 to rt7 patch. Can you confirm?
> 
> If so, which recent kernel should I use in order to get smp and rt
> running without problems?
> 
This simple patch should solve the above problem. Seeing the config, its
seems to be a 32bit box.


It convert tlbstate_lock spin_lock to raw_spin_lock.
Preemption has already been disabled, hence this change shouldn't
affect latency numbers.


Signed-Off-By: Chirag <chirag@linux.vnet.ibm.com>


---
 arch/x86/kernel/tlb_32.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.26.5-rt8/arch/x86/kernel/tlb_32.c
===================================================================
--- linux-2.6.26.5-rt8.orig/arch/x86/kernel/tlb_32.c	2008-09-09 21:45:25.000000000 -0400
+++ linux-2.6.26.5-rt8/arch/x86/kernel/tlb_32.c	2008-09-09 21:50:42.000000000 -0400
@@ -23,7 +23,7 @@ DEFINE_PER_CPU(struct tlb_state, cpu_tlb
 static cpumask_t flush_cpumask;
 static struct mm_struct *flush_mm;
 static unsigned long flush_va;
-static DEFINE_SPINLOCK(tlbstate_lock);
+static DEFINE_RAW_SPINLOCK(tlbstate_lock);
 
 /*
  * We cannot call mmdrop() because we are in interrupt context,
